File: src/app/common/movie.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { MovieResponse, Movie } from './movie/movie';

@Injectable()
export class MovieService {
  constructor(private httpClient: HttpClient) {}

  search(query: string, page: string) {
    const params = new HttpParams()
      .set('query', query.trim())
      .set('page', page);
    return this.httpClient.get('/movie/search', { params });
  }

  upcoming(page: string) {
    const params = new HttpParams().set('page', page);
    return this.httpClient.get<MovieResponse>('/movie/forthcoming', { params });
  }

  popular(page: string) {
    const params = new HttpParams().set('page', page);
    return this.httpClient.get<MovieResponse>('/movie/trending', { params });
  }

  recommend(movie: Movie) {
    return this.httpClient.post('/movie/my-recommendations', movie);
  }

  unrecommend(movie: Movie) {
    return this.httpClient.delete(`/movie/my-recommendations/${movie.id}`);
  }

  getMovies(page: string) {
    const params = new HttpParams().set('page', page);
    return this.httpClient.get<MovieResponse>('/movie/my-recommendations', {
      params
    });
  }

  recommendations(movie: Movie) {
    const params = new HttpParams().set('id', movie.id + '');
    return this.httpClient.get<MovieResponse>('/movie/recommended', { params });
  }

  toggleRecommendation(movies: Movie[], movie: Movie) {
    if (movies) {
      const index = movies.findIndex(m => m.id === movie.id);
      if (index > -1) {
        movies[index] = movie;
      }
    }
  }
}

File: Server/src/main/java/com/stackroute/movieapp/recommendations/control/MovieService.java
package com.stackroute.movieapp.recommendations.control;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.MediaType;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.convert.converter.Converter;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import com.stackroute.movieapp.Config;
import com.stackroute.movieapp.recommendations.entity.Movie;
import com.stackroute.movieapp.recommendations.entity.MovieResponse;
import com.stackroute.movieapp.tmdb.search.dto.SearchResponse;

@Service
public class MovieService {

	private static final String API_KEY = "api_key";

	@Autowired
	private Config properties;

	@Autowired
	private Converter<SearchResponse, MovieResponse> converter;

	@Autowired
	private MovieRepository movieRepository;

	private Client client;

	public MovieService() {
		client = ClientBuilder.newClient();
	}

	public MovieResponse search(String query, String page) {
		WebTarget target = client.target(properties.getBaseUrl()).path(properties.getSearchPath());
		SearchResponse response = target.queryParam(API_KEY, properties.getKey()).queryParam("query", query)
				.queryParam("page", page).request(MediaType.APPLICATION_JSON).get(SearchResponse.class);

		return converter.convert(response);
	}

	public MovieResponse forthcoming(String page) {
		WebTarget target = client.target(properties.getBaseUrl()).path(properties.getForthcomingPath());
		SearchResponse response = target.queryParam(API_KEY, properties.getKey()).queryParam("page", page)
				.request(MediaType.APPLICATION_JSON).get(SearchResponse.class);

		return converter.convert(response);
	}

	public MovieResponse trending(String page) {
		WebTarget target = client.target(properties.getBaseUrl()).path(properties.getTrendingPath());
		SearchResponse response = target.queryParam(API_KEY, properties.getKey()).queryParam("page", page)
				.request(MediaType.APPLICATION_JSON).get(SearchResponse.class);

		return converter.convert(response);
	}

	public MovieResponse recommended(Long id, String page) {
		WebTarget target = client.target(properties.getBaseUrl()).path(properties.getRecommendedPath())
				.resolveTemplate("id", id);
		SearchResponse response = target.queryParam(API_KEY, properties.getKey()).queryParam("page", page)
				.request(MediaType.APPLICATION_JSON).get(SearchResponse.class);

		return converter.convert(response);
	}

	public MovieResponse getMyRecommendations(int page) {
		Page<Movie> movies = movieRepository.findAll(new PageRequest(page - 1, 20));
		MovieResponse response = new MovieResponse();
		response.setMovies(movies.getContent());
		response.setPage(movies.getNumber());
		response.setTotalPages(movies.getTotalPages());
		response.setTotalResults(Integer.valueOf(movies.getTotalElements() + ""));
		return response;
	}

	public MovieResponse save(Movie movie) {
		Movie savedMovie = movieRepository.save(movie);
		MovieResponse response = new MovieResponse();
		response.getMovies().add(savedMovie);
		return response;
	}

	public MovieResponse delete(Long id) {
		movieRepository.delete(id);
		MovieResponse response = new MovieResponse();
		response.setStatus("success");
		return response;
	}
}
